src/ApimEventProcessor/MoesifHttpMessageProcessor.cs

```diff
@@ -28,52 +34,96 @@ public MoesifHttpMessageProcessor(ILogger logger)
 
         public async Task ProcessHttpMessage(HttpMessage message)
         {
-            if (message.IsRequest)
-            {
+            // message will probably contain either HttpRequestMessage or HttpResponseMessage
+            // So we cache both request and response and cache them. 
+            // Note, response message might be processed before request
+            if (message.HttpRequestMessage != null){
                 message.HttpRequestMessage.Properties.Add(RequestTimeName, DateTime.UtcNow);
-                return;
+                requestsCache.TryAdd(message.MessageId, message);
             }
-            
+            if (message.HttpResponseMessage != null){
+                responsesCache.TryAdd(message.MessageId, message);
+            }
+            await SendCompletedMessagesToMoesif();
+        }
+
+        /*
+        From requestCache and responseCache, find all messages that have request and response
+        Send them to Moesif asynchronously.
+        */
+        public async Task SendCompletedMessagesToMoesif()
+        {
+            var completedMessages = RemoveCompletedMessages();
+            _Logger.LogDebug("Sending completed Messages to Moesif. Count: " + completedMessages.Count);
+            foreach(Guid messageId in completedMessages.Keys)
+            {
+                var kv = completedMessages[messageId];
+                var moesifEvent = await BuildMoesifEvent(kv.Key, kv.Value);
+                var response = await _MoesifClient.Api.CreateEventAsync(moesifEvent);
+            }
```

Comment:
I know in this case, since SendCompletedMessagesToMoesif is called every time,  here foreach runs only once. However, we could improve that may be later to call SendCompletedMessagesToMoesif either a batch fills or every x seconds, so we drain events.

Atleast in line 58, you have array of messages, can we use CreateEventAsyncBatch assuming _MoesifClient.Api has that support.

Reply:
Sounds good. I will setup the scenario and test it tomorrow.